### ISSUE TITLE
EunomiaDatasets format

### ISSUE THREAD

**Issue by ablack3** [OHDSI/Eunomia#50]
I've built some additional Eunomia datasets available [here]( https://drive.google.com/drive/folders/1SROTshSBoTYq7MsRtZjYhYwQhmiOHPy5?usp=sharing) in duckdb v0.8 format.

I was wondering if we can use parquet files instead of csv since parquet is smaller than csv.

I was also wondering if I can include a full vocabulary as the full vocab tables are very helpful for examples.

All of these cdms use the same vocab so it would be more efficient to store the vocab tables just once and allow eunomia to use two seperate links, one for vocab tables and one for the other cdm tables.

What do you all think?

@fdefalco, @schuemie

**Comment by fdefalco** [comment]
I'm open to switching to parquet.  I would like to test/benchmark the impact of the switch on file size and load performance time.

We haven't included a full vocabulary in the past because of licensing of some vocabularies being different across organizations.  The size of the overall vocabulary is also somewhat cumbersome to include in the sample as it would account for 95% of the data size if included in full.

My desire was to have an API for Athena that could be used to construct a vocabulary to support a particular data source in a meaningful way, but that has not been implemented.

I know that some of our samples contain different vocabulary content because we used some pruning techniques to only include vocabulary content that applied to the content of the clinical tables. In that case, a single, complete vocabulary might also be cumbersome.  That depends on the loading and query time of whatever technologies we are using for storage and loading.  Perhaps the switch to parquet alleviates those concerns?

**Comment by schuemie** [comment]
In my experience Parquet is really efficient, and at least in Python you can directly run DuckDB of Parquet files (which is also really performant). This is certainly an interesting area to explore.

**Comment by ablack3** [comment]
The full vocab is soooo nice though. And it's optional. The default Eunomia is still the same as it has been. But the user could choose to download a larger eunomia cdm with a full vocab. I don't know much about licensing issues but my understanding is that there is some issue with the CPT code descriptions...

Yea a more complete vocabulary is very nice for testing and examples. The current Eunomia dataset does not even have domain_id filled in which caused an issue for me in testing. One possible approach is for me to experiment a bit in CDMConnector and the contribute things that seem to work well into Eunomia since Eunomia has more reverse dependencies.

In the example datasets I created the vocab is the bulk of the data so it would be nice to not repeat that multiple times in the github repo.

**Comment by ablack3** [comment]
I agree an Athena API is needed. I do know one place to download full vocabs from a public s3 bucket though... https://github.com/OHDSI/OHDSIonAWS/blob/0ecc5f2af0bd2b487804612b82f86a1b2d1577bf/datasources/CMSDESynPUF100k.sql#L25C31-L25C76

So it seems like Eunomia should be ok too then. 😄

**Comment by fdefalco** [comment]
My suggestion is to confirm with the vocabulary team that we're allowed to openly distribute the complete vocabulary before adding it to any particular package.  This might also be an opportunity to version the vocabulary releases, which wouldn't necessarily go in the Eunomia package, but we could provide a separate repository and then enable downloads much like we do with Eunomia. 

The sample data sets typically have a very small subset of concept identifiers in them, what benefit is there of having the complete vocabulary in those cases?

**Comment by ablack3** [comment]
> The sample data sets typically have a very small subset of concept identifiers in them, what benefit is there of having the complete vocabulary in those cases?

Keep in mind that many people just getting in to OHDSI have no access to any OMOP CDM data. Hard to imagine for the veterans I know 😄  But for newcomers, being able to very quickly start using an OMOP CDM is a game-changer for onboarding similar to other efforts like OHDSI in a Box and OHDSI on AWS which both provide full CDM datasets with full vocabularies.

An incomplete CDM is a very strange CDM to use because you don't really know what can be done and what cannot. Even after using Eunomia for years I just recently realized that I could not rely on concept.domain_id in Eunomia because it is not populated. So I need to create something else to test any code that relies on domain_id. I could hardly find any duplicated concepts in Eunomia (i.e. a person who has the same concept more than once).

Eunomia only supports very specific cohorts. A small-ish but still complete CDM like the ~8GB covid19_200k synthea CDM I created for the Oxford summer school last week allows newcomers to run entire study on somewhat realistic data. My hope/plan is to push this even further by making the synthetic data in Eunomia more closely aligned to the real data that is not public and newcomers, independent researchers, and people working at small companies don't generally have access to.

**Comment by ablack3** [comment]
@cgreich. Welcome to this conversation on Eunomia!

Eunomia is an important tool for training newcomers to OHDSI and testing our software. However the current Eunomia dataset is limited by subset of the vocabulary. 

Other tools used in OHDSI for training such as OHDSI-in-a-box and OHDSI-on-AWS provide full vocabulary tables to their users (not including any of the licensed vocabs of course).

Our question for you is "Can we include a full vocabulary in the example synthetic CDMs in Eunomia"? If not, why? Do the same restrictions apply to OHDSI-in-box and OHDSI-on-AWS?

Thanks!

**Comment by fdefalco** [comment]
> I'm not really sure of all the implications of including a vocab with just the subset of concepts that occur in the CDM data. Certainly if you do this you won't get correct answers to queries like "What are all the descendants of concept X?" so it won't be a good tool to train people to use and experiment with the vocab tables.

Actually, for the sample data that I had created, the full hierarchy for concepts that occur in the event tables were included. But yes, if you attempted that type of query on the sample data set for a concept not represented you would get no results.

> One exercise we did last week was to build a concept set for a clinical idea and I think for that exercise it's helpful to have a complete vocabulary even if those concepts don't all occur in the data.

Isn't the demo ATLAS a good tool for building a concept set?  Or were you focusing on how to write SQL to perform this task? 

> This reminds me of similar discussion I've had pondering if a CDM is missing tables or fields should it still be considered a CDM? Why not leave off the irrelevant tables? In practice I think that in the wild CDM tables or columns can be entirely missing and the user has to adapt and decide if it is still a "valid cdm".

I think this conflates two different ideas.  It is one thing to leave tables and/or columns out of the CDM which would modify the database's structure.  It is something different to limit the content in the tables.  

> An incomplete CDM is a very strange CDM to use because you don't really know what you're missing. "Is my query/test giving an unexpected result because my vocab is incomplete?" is a constant question to consider. Even after using Eunomia for years I just recently realized that I could not rely on concept.domain_id in Eunomia because it is not populated until it caused a test to fail.

Also at this point I think we need to clarify which data set you are describing as Eunomia is no longer a single data set, but a utility to pull a particular data set.  So if one of the sample data sets does not have a populated concept.domain_id we should fix that as an issue.

> Eunomia only supports very specific cohorts. 

Again, this will vary by the data set you are pulling.

> Imagine if Eunomia had a function you could point at a CDM and optionally a generated cohort in the CDM that would output a synthea json spec representing the state transitions (codes and their probabilities) that synthea could use to generate synthetic data matching that patient population and CDM database. Doesn't seem impossible.

This is a very interested idea and one that I have discussed with several people over the years.  There are companies that have proprietary technology to create synthetic data from real data with some level of complete statistical compatibility.  I have written some code that starts to do what you are describing with the synthea modules generated by CDM driven information and I think there is a lot of potential there however I do not think it should be part of Eunomia.  Eunomia should remain a way to access sample data sets.  Developing methods to generate modules from CDM data to output Synthea modules is a project I would definitely like to be a part of if that is the goal.  One question that has been raised by this however, is for the underlying data set that you would be using to 'train' your module, at what point do you break the legal / license obligations by publishing data.  For example if I take a licensed data set, randomize one variable and then publish it, I would be in breach of contract.  Clearly these are very different things, but it is a consideration that must be explored to determine when a process would result an a 'digital twin' representation of a database and be akin to publishing data that is otherwise unauthorized to be published.

I think an interim step that could be taken is instead to increase concept representation in the synthea generated data but writing additional modules or authoring changes to existing modules.  For example, we can see from that PNAS publication that the set of drugs used to treat hypertension is much broader than the data that the modules currently generate.  

https://www.pnas.org/doi/10.1073/pnas.1510502113

Improving modules so they more accurately reflect even published findings would be a valuable contribution.

**Comment by ablack3** [comment]
Thanks for that feedback @fdefalco!

In my examples when I refer the "the Eunomia dataset" I'm referring to the gibleed dataset that is currently the only dataset provided by latest release of the Eunomia package. It looks like concept.domain_id is there so I must have made another mistake in my code somewhere.

I'm not sure how limiting the number of concepts in the concept table really alleviates vocabulary licensing/redistribution concerns though. Is 50% of the vocabulary ok to distribute? What about 90%?

But really the issue is one of closure. The OMOP vocabulary is a closed set of relationships meaning that I can move around through the relationships and stay inside the set of concepts. By contrast the limited vocabulary in the GiBleed Eunomia data set is not closed.  Here is a reprex showing that many related concepts are not in the concept table. So if we use a subset of the vocabulary I think it should at least retain the closure property. 

``` r
cd <- Eunomia::getEunomiaConnectionDetails()

con <- DatabaseConnector::connect(cd)
#> Connecting using SQLite driver

library(DatabaseConnector)

df <- querySql(con, 
         "select c1.concept_id, c2.concept_id as related_concept_id
         from concept c1
         left join concept_relationship cr on c1.concept_id = cr.concept_id_1
         left join concept c2 on cr.concept_id_2 = c2.concept_id
         ")

df
#>     CONCEPT_ID RELATED_CONCEPT_ID
#> 1     35208414             192671
#> 2     35208414            1569708
#> 3      1118088           44923712
#> 4     40213201                 NA
#> 5      1557272                 NA
#> 6      4336464                 NA
#> 7      4295880                 NA
#> 8      3020630                 NA
#> 9     19129655                 NA
#> 10    44923712            1118088
#> 11     1569708           35208414
#> 12    40213216                 NA
#> 13      708298                 NA
#> 14     4301346                 NA
#> 15     4246502                 NA
#> 16     3020655                 NA
#> 17     4116491                 NA
#> 18    45011828           40162359
#> 19    40162359           45011828
#> 20    40213227                 NA
#> 21      701322                 NA
#> 22     3002385                 NA
#> 23     4199276                 NA
#> 24     3005136                 NA
#> 25     4113008                 NA
#> 26    40213230                 NA
#> 27      723013                 NA
#> 28     3001247                 NA
#> 29     4234705                 NA
#> 30     3004501                 NA
#> 31     4156265                 NA
#> 32    40213160                 NA
#> 33     1129625                 NA
#> 34     3036780                 NA
#> 35     4238715                 NA
#> 36     3006734                 NA
#> 37     4155034                 NA
#> 38    40213180                 NA
#> 39     1149196                 NA
#> 40     3006451                 NA
#> 41     4198190                 NA
#> 42     3027970                 NA
#> 43     4109685                 NA
#> 44    40213251                 NA
#> 45     1149380                 NA
#> 46     3006923                 NA
#> 47     4202832                 NA
#> 48     3023103                 NA
#> 49     4048169                 NA
#> 50    40213260                 NA
#> 51     1150770                 NA
#> 52     3007461                 NA
#> 53     4151422                 NA
#> 54     3023430                 NA
#> 55     4240902                 NA
#> 56    40213281                 NA
#> 57     1150836                 NA
#> 58     3008598                 NA
#> 59     4163872                 NA
#> 60     3027231                 NA
#> 61     4052083                 NA
#> 62    40213296                 NA
#> 63     1124300                 NA
#> 64     3009201                 NA
#> 65     4163951                 NA
#> 66     3026361                 NA
#> 67     4012477                 NA
#> 68    40213299                 NA
#> 69     1153428                 NA
#> 70     3009542                 NA
#> 71     4163971                 NA
#> 72     3024731                 NA
#> 73     4047491                 NA
#> 74    40213306                 NA
#> 75     1107830                 NA
#> 76     3010813                 NA
#> 77     4094814                 NA
#> 78     3024561                 NA
#> 79     4078793                 NA
#> 80    40213314                 NA
#> 81     1110410                 NA
#> 82     3012494                 NA
#> 83     4048695                 NA
#> 84    40486433                 NA
#> 85     4043071                 NA
#> 86    40213320                 NA
#> 87     1115008                 NA
#> 88     3013682                 NA
#> 89     4051466                 NA
#> 90    40769179                 NA
#> 91     4037675                 NA
#> 92    40213198                 NA
#> 93      753626                 NA
#> 94     3011505                 NA
#> 95     4142905                 NA
#> 96    40769184                 NA
#> 97     4275113                 NA
#> 98    40213154                 NA
#> 99     1728416                 NA
#> 100     933724                 NA
#> 101    4144583                 NA
#> 102   40769189                 NA
#> 103    4108705                 NA
#> 104   42902245                 NA
#> 105     738818                 NA
#> 106     926487                 NA
#> 107    4296204                 NA
#> 108   40769194                 NA
#> 109    4242997                 NA
#> 110   40221415                 NA
#> 111    1124957                 NA
#> 112    1102527                 NA
#> 113    4296205                 NA
#> 114    3035995                 NA
#> 115    4293740                 NA
#> 116   43012433                 NA
#> 117    1125315                 NA
#> 118     782043                 NA
#> 119    4294548                 NA
#> 120    1310034                 NA
#> 121    4243062                 NA
#> 122   40169216                 NA
#> 123    1125443                 NA
#> 124     791967                 NA
#> 125    4278672                 NA
#> 126   40173590                 NA
#> 127    4117038                 NA
#> 128   40243436                 NA
#> 129    1118084                 NA
#> 130     920293                 NA
#> 131    4237458                 NA
#> 132   42707627                 NA
#> 133    4107731                 NA
#> 134   40224132                 NA
#> 135    1119510                 NA
#> 136   19010482                 NA
#> 137    4283893                 NA
#> 138    4216130                 NA
#> 139    4218389                 NA
#> 140   40224805                 NA
#> 141    1137529                 NA
#> 142   19078219                 NA
#> 143    4280726                 NA
#> 144    4306014                 NA
#> 145    4310024                 NA
#> 146    1154615                 NA
#> 147     757627                 NA
#> 148   19054876                 NA
#> 149    4285898                 NA
#> 150   19133679                 NA
#> 151    4146173                 NA
#> 152   40220386                 NA
#> 153     740275                 NA
#> 154   19074841                 NA
#> 155    4149245                 NA
#> 156   44805732                 NA
#> 157    4266809                 NA
#> 158   40223774                 NA
#> 159    1154029                 NA
#> 160    1127078                 NA
#> 161    4152936                 NA
#> 162   42535580                 NA
#> 163    4230399                 NA
#> 164   40171902                 NA
#> 165    1154343                 NA
#> 166    1127433                 NA
#> 167    4166224                 NA
#> 168   21494966                 NA
#> 169    4304110                 NA
#> 170    1718412                 NA
#> 171       9201                 NA
#> 172     920300                 NA
#> 173    3014599                 NA
#> 174   45769019                 NA
#> 175    4169137                 NA
#> 176   40174880                 NA
#> 177       9202                 NA
#> 178     920334                 NA
#> 179    1332419                 NA
#> 180   45766064                 NA
#> 181    4219399                 NA
#> 182   40222846                 NA
#> 183       9203                 NA
#> 184   19019273                 NA
#> 185   19057271                 NA
#> 186   46221626                 NA
#> 187    4314290                 NA
#> 188   40223768                 NA
#> 189          0                 NA
#> 190   19007333                 NA
#> 191    1589505                 NA
#> 192   45892894                 NA
#> 193    4133840                 NA
#> 194   40223821                 NA
#> 195     433644                 NA
#> 196   19019979                 NA
#> 197     705944                 NA
#> 198   19112599                 NA
#> 199    4252419                 NA
#> 200   40223834                 NA
#> 201     442116                 NA
#> 202   19074843                 NA
#> 203     715997                 NA
#> 204   19077344                 NA
#> 205    4218408                 NA
#> 206   40220960                 NA
#> 207     438614                 NA
#> 208   19075001                 NA
#> 209    1322184                 NA
#> 210   19078924                 NA
#> 211    4170947                 NA
#> 212   43012036                 NA
#> 213     198809                 NA
#> 214   19073183                 NA
#> 215    1367571                 NA
#> 216   19076374                 NA
#> 217    4230209                 NA
#> 218    1594382                 NA
#> 219     195588                 NA
#> 220   19073188                 NA
#> 221    1500211                 NA
#> 222   19079524                 NA
#> 223    4337138                 NA
#> 224   43012754                 NA
#> 225     378001                 NA
#> 226   19075601                 NA
#> 227    1309944                 NA
#> 228   19077572                 NA
#> 229    4337611                 NA
#> 230   40229134                 NA
#> 231     378419                 NA
#> 232   19059056                 NA
#> 233    1305058                 NA
#> 234   19029476                 NA
#> 235    4168004                 NA
#> 236   40236824                 NA
#> 237     317009                 NA
#> 238   19078461                 NA
#> 239    1347450                 NA
#> 240   19045231                 NA
#> 241   19016749                 NA
#> 242   40236446                 NA
#> 243     317576                 NA
#> 244    4323902                 NA
#> 245    1310149                 NA
#> 246   19074679                 NA
#> 247   19003953                 NA
#> 248    1713671                 NA
#> 249     258780                 NA
#> 250    4324181                 NA
#> 251    1326303                 NA
#> 252    1729720                 NA
#> 253   19006318                 NA
#> 254     964261                 NA
#> 255     260139                 NA
#> 256   40481087                 NA
#> 257    1332418                 NA
#> 258     742185                 NA
#> 259   44816277                 NA
#> 260   40231925                 NA
#> 261     133834                 NA
#> 262    3006906                 NA
#> 263    1361711                 NA
#> 264    1112807                 NA
#> 265   35605482                 NA
#> 266    1115171                 NA
#> 267     134438                 NA
#> 268    3045716                 NA
#> 269     372328                 NA
#> 270     967823                 NA
#> 271   35605858                 NA
#> 272   40163554                 NA
#> 273      78272                 NA
#> 274    3028833                 NA
#> 275     375671                 NA
#> 276    1778162                 NA
#> 277   40766240                 NA
#> 278   40162522                 NA
#> 279      80809                 NA
#> 280   40232448                 NA
#> 281     313217                 NA
#> 282    1512674                 NA
#> 283   46275444                 NA
#> 284    1519937                 NA
#> 285      81151                 NA
#> 286     782047                 NA
#> 287     321042                 NA
#> 288    1177480                 NA
#> 289   46275916                 NA
#> 290    1551192                 NA
#> 291      28060                 NA
#> 292    4048171                 NA
#> 293     261325                 NA
#> 294    1192710                 NA
#> 295    1396131                 NA
#> 296    1539411                 NA
#> 297      30753                 NA
#> 298    4035415                 NA
#> 299     441267                 NA
#> 300    1343916                 NA
#> 301    1307863                 NA
#> 302    1545959                 NA
#> 303     196456                 NA
#> 304    4112343                 NA
#> 305     439777                 NA
#> 306   19039298                 NA
#> 307    1308738                 NA
#> 308    1150771                 NA
#> 309      80180                 NA
#> 310     914335                 NA
#> 311     380378                 NA
#> 312   19059528                 NA
#> 313    1501700                 NA
#> 314    1150837                 NA
#> 315      80502                 NA
#> 316     836654                 NA
#> 317     381316                 NA
#> 318   40479422                 NA
#> 319    1503184                 NA
#> 320    1107882                 NA
#> 321      81893                 NA
#> 322     836659                 NA
#> 323     433125                 NA
#> 324   40479768                 NA
#> 325   40480160                 NA
#> 326    1595799                 NA
#> 327     198199                 NA
#> 328    3000963                 NA
#> 329     440086                 NA
#> 330   40478878                 NA
#> 331    4202451                 NA
#> 332    1717327                 NA
#> 333     255848                 NA
#> 334    3001488                 NA
#> 335     440448                 NA
#> 336    3006322                 NA
#> 337   19089061                 NA
#> 338    1738521                 NA
#> 339     257012                 NA
#> 340    3019550                 NA
#> 341    4001336                 NA
#> 342    4125906                 NA
#> 343   19115351                 NA
#> 344     975125                 NA
#> 345     192671           35208414
#> 346    3015632                 NA
#> 347    4029498                 NA
#> 348    4132546                 NA
#> 349   19128263                 NA
#> 350    1000560                 NA
#> 351     140673                 NA
#> 352    3015076                 NA
#> 353    4027663                 NA
#> 354    4174876                 NA
#> 355   19127904                 NA
#> 356    1741122                 NA
#> 357    4066995                 NA
#> 358    3016723                 NA
#> 359    4043241                 NA
#> 360    4139681                 NA
#> 361   19127918                 NA
#> 362    1759842                 NA
#> 363    4059173                 NA
#> 364    3021226                 NA
#> 365    4084167                 NA
#> 366   19018935                 NA
#> 367   19127922                 NA
#> 368    1713332                 NA
#> 369    4056621                 NA
#> 370    3019406                 NA
#> 371    4134304                 NA
#> 372   19023398                 NA
#> 373   19128009                 NA
#> 374    1746114                 NA
#> 375    4299128                 NA
#> 376    3014576                 NA
#> 377    4015189                 NA
#> 378   19023423                 NA
#> 379   19128022                 NA
#> 380    1768849                 NA
#> 381    4329847                 NA
#> 382    3013721                 NA
#> 383    4010253                 NA
#> 384   19008572                 NA
#> 385   19127890                 NA
#> 386    1174888                 NA
#> 387     439406                 NA
#> 388    3000876                 NA
#> 389    4035793                 NA
#> 390   19002770                 NA
#> 391   19123231                 NA
#> 392    1515774                 NA
#> 393     443279                 NA
#> 394    3051031                 NA
#> 395    4186930                 NA
#> 396   19132304                 NA
#> 397   44782520                 NA
#> 398    1518198                 NA
#> 399    4102123                 NA
#> 400    4326177                 NA
#> 401    4191853                 NA
#> 402   19134077                 NA
#> 403   40758406                 NA
#> 404    1519936                 NA
#> 405    4302207                 NA
#> 406    4323208                 NA
#> 407    4181917                 NA
#> 408   19128065                 NA
#> 409   46235214                 NA
#> 410    1521369                 NA
#> 411    4024958                 NA
#> 412   40492359                 NA
#> 413    4187458                 NA
#> 414   19129144                 NA
#> 415   37017711                 NA
#> 416    1551099                 NA
#> 417    4024289                 NA
#> 418   40169281                 NA
#> 419    4145310                 NA
#> 420   19133768                 NA
#> 421   46233710                 NA
#> 422    1539403                 NA
#> 423    4057420                 NA
#> 424   40228214                 NA
#> 425    4197460                 NA
#> 426   19133873                 NA
#> 427   46233988                 NA
#> 428    1545958                 NA
#> 429    4059978                 NA
#> 430   40228230                 NA
#> 431    4195803                 NA
#> 432   19133905                 NA
#> 433   44783196                 NA
#> 434    1548195                 NA
#> 435    4062501                 NA
#> 436   40165015                 NA
#> 437    4330583                 NA
#> 438   19125062                 NA
#> 439    1549786                 NA
#> 440    4058899                 NA
#> 441    4163874                 NA
#> 442    3049273                 NA
#> 443   19126352                 NA
#> 444       8507                 NA
#> 445       8532                 NA
disconnect(con)
```

<sup>Created on 2023-06-29 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

We can move the discussion about the `train_synthea(cdm) -> json::synthea_module` function to another thread. I'll invite you to a project in-progress https://github.com/ablack3/SyntheaCdmFactory based largely on your work on ETL-Synthea. I'd love to collaborate on this. I just run the ETL in duckdb instead of postgres which makes it easier.

**Comment by ablack3** [comment]
Here I tried (in a short amount of time) to find all concepts in the gibleed Eunomia dataset and then check if they were in the concept table. Maybe I made a mistake but it looks to me like there are many concept in the gibleed vocabulary that are not in the concept table. I think creating a "closed" subset of the vocabulary is not a trivial task and you need to make decisions about what relationships to include and which not to include. To me it's much easier just to use the whole vocabulary as it is released. Plus then you can run Patrick's SQL queries that create cool messages.

``` r
cd <- Eunomia::getEunomiaConnectionDetails()

con <- DatabaseConnector::connect(cd)
#> Connecting using SQLite driver

DBI::dbListTables(con)
#>  [1] "care_site"             "cdm_source"            "cohort"               
#>  [4] "cohort_attribute"      "concept"               "concept_ancestor"     
#>  [7] "concept_class"         "concept_relationship"  "concept_synonym"      
#> [10] "condition_era"         "condition_occurrence"  "cost"                 
#> [13] "death"                 "device_exposure"       "domain"               
#> [16] "dose_era"              "drug_era"              "drug_exposure"        
#> [19] "drug_strength"         "fact_relationship"     "location"             
#> [22] "measurement"           "metadata"              "note"                 
#> [25] "note_nlp"              "observation"           "observation_period"   
#> [28] "payer_plan_period"     "person"                "procedure_occurrence" 
#> [31] "provider"              "relationship"          "source_to_concept_map"
#> [34] "specimen"              "visit_detail"          "visit_occurrence"     
#> [37] "vocabulary"
library(DatabaseConnector)

df <- querySql(con, 
  "
  with cte as(
    select ancestor_concept_id as id from concept_ancestor
    union
    select descendant_concept_id as id from concept_ancestor
    union
    select concept_id_1 as id from concept_relationship
    union 
    select concept_id_2 as id from concept_relationship
    union
    select vocabulary_concept_id as id from vocabulary
  ) 
  select distinct id, concept_id from cte a
  left join concept b on a.id = b.concept_id
")

head(df, 20)
#>       ID CONCEPT_ID
#> 1      0          0
#> 2    204         NA
#> 3    231         NA
#> 4    232         NA
#> 5    236         NA
#> 6    238         NA
#> 7    243         NA
#> 8    245         NA
#> 9    252         NA
#> 10  5029         NA
#> 11  5045         NA
#> 12  5046         NA
#> 13  5047         NA
#> 14  9201       9201
#> 15  9202       9202
#> 16  9203       9203
#> 17 24818         NA
#> 18 25297         NA
#> 19 28060      28060
#> 20 30753      30753
dplyr::count(df, is.na(CONCEPT_ID))
#>   is.na(CONCEPT_ID)     n
#> 1             FALSE   442
#> 2              TRUE 39418
disconnect(con)
```

<sup>Created on 2023-06-29 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by ablack3** [comment]
> Isn't the demo ATLAS a good tool for building a concept set? Or were you focusing on how to write SQL to perform this task?

Atlas is a great tool for building concept sets. But yes our workshop was focused on doing the same thing from R using the Darwin CodelistGenerator package which writes the SQL for you and have various options for searching the vocab. The nice thing about this is that you can fairly easily take that code list and convert it to a generated cohort _on your data_ using Capr. Of course Atlas makes this easy as well but I met several people working at organizations that have OMOP data but don't have Atlas because it's actually not very easy to set up. I have personal experience with it taking more than two years to deploy Atlas and it sounds like that experience isn't unique. Lots of hurtles, not just technical.

**Comment by fdefalco** [comment]
Part of the limitation on the Eunomia vocabulary was based on size, we pruned until we got under the 5mb limit that was enforced by CRAN.  The other difference you might be seeing is based on the version of the vocabulary, as that was created years ago.  

To be clear, I'm not opposed to providing an easier way to obtain a copy of the vocabulary.  Somewhere along the way I understood we were not allowed to openly distribute the entire vocabulary.  If @cgreich can approve OHDSI's ability to openly distribute all of the vocabulary, then having it available as parquet files as a dataset in a vocabulary repository would be great.
